File: Codes/Algorithm/27.remove-element.cpp
/*
 * @lc app=leetcode id=27 lang=cpp
 *
 * [27] Remove Element
 */

// @lc code=start
class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        int slow_pointer = 0;
        for(int fast_pointer=0; fast_pointer < nums.size(); fast_pointer++) {
            if (nums[fast_pointer] != val) {
                nums[slow_pointer++] = nums[fast_pointer]; 
            }
        }
        return slow_pointer;
    }
};
// @lc code=end

File: Codes/Algorithm/27.remove-element.py
#
# @lc app=leetcode id=27 lang=python
#
# [27] Remove Element
#

# @lc code=start
class Solution(object):
    def removeElement(self, nums, val):
        """
        :type nums: List[int]
        :type val: int
        :rtype: int
        """
        
        '''
        Dual pointer:
        Two pointer fast and slow, for each elements, if the element 
        not equal to the target, the element of slow pointer is re-writted
        by the fast pointer element. Then slow pointer move forward. 
        Whatever the element is, fast pointer always move forward.
        '''
        fast = 0
        slow = 0
        while fast < len(nums):
            if nums[fast] != val:
                nums[slow] = nums[fast]
                slow += 1
            fast += 1
        return slow

# @lc code=end

